# Functional Annotation Guide

**Collection of command line protocols of functional annotation tools in plant genomics**

## Overview

Many tools are available for various tasks in plant genomics. Finding the right tool for a certain purpose can be challenging. This repository provides an overview of recommended tools specifically for functional annotation in plant genomics. This documentation is a guide elucidating step-by-step usage of some widely used functional annotation tools in plant genomics. The tools covered range from user-friendly databases to advanced command-line tools. This will be of use to biologists interested in functional annotation of genes of interest in their own research. Each of the commands and usage instructions here are based on an example file named 'sample' and need to adapted based on the user's own use-case. If the user is not familiar with linux environment and packages for software installation, it is first recommended to read these relevant guidelines specified in https://github.com/PuckerLab/PlantGenomicsGuide.


## Functional annotation database tools

### InterProScan5

### Mercator4

### KEGG

### BLAST2GO


## Ortholog finding tools

### OrthoFinder2

### SHOOT

### FASTOMA

## Sequence similarity based tools (aligners)

### BLAST

### DIAMOND

### MMSeqs

### MAFFT

### MUSCLE5

## Phylogeny and visualization tools

### IQ-TREE3

### FastTree

### iTOL

## Synteny analysis tools

### JCVI/ MCScan

### TBtools-II

### TOGA

### Orthorefine

### SOI

## Protein structure-based annotation tools

### Dali

### Foldseek

## Expression analysis-based annotation tools

### GENIE3

### WGCNA

## Annotation tools employing combined approaches

### CoExpPhylo

### KIPEs

### DupyliCate

## References



If you have questions about plant genomics that were not answered by any of these resources, please feel free to get in touch with the Plant Biotechnology and Biotechnology research group at the University of Bonn.
